### PR TITLE
Ryt tmp

### PR DESCRIPTION
**Pull request by declmal**
Thanks for contributing to TVM!   Please refer to guideline https://docs.tvm.ai/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/dmlc/tvm/blob/master/CONTRIBUTORS.md#reviewers).

### CODE REVIEW

File: cvm/quantization/from_tensorflow.py [hunk 1]
```diff
@@ -592,10 +633,15 @@ def convert_tfnode(tfnode, graph, params, infer_shapes, logger=logging):
             elif k not in ('dtype', '_output_shapes', '_class'):
                 raise NotImplementedError \
                     ("Other attributes for a Const(param) Node {} ? .".format(k))
-    elif op_name in ['Placeholder', 'PlaceholderWithDefault']:
+    elif op_name in ['Placeholder', 'PlaceholderWithDefault', 'FIFOQueueV2']:
         input_shape = \
             tensor_util.tensor_shape_proto_to_list(attr['shape'].shape)
-        dtype = tensor_util.tensor_type_to_numpy(attr['dtype'].type)
+        if not input_shape:
+            input_shape = default_input_shape
```
Comment (wlt-cortex) [inline cvm/quantization/from_tensorflow.py]: Cannot the input_shape be extracted from the attr['shape'], or the shape is not indicated in the TensorFlow model?

I suggest that the input shape should be passed through the parameters list with user-settings, instead of setting default shape.

File: cvm/quantization/from_tensorflow.py [hunk 2]
```diff
@@ -592,10 +633,15 @@ def convert_tfnode(tfnode, graph, params, infer_shapes, logger=logging):
...
+        try:
+            dtype = tensor_util.tensor_type_to_numpy(attr['dtype'].type)
+        except TypeError:
+            dtype = tensor_util.tensor_type_to_numpy(default_tf_dtype)
```
Comment (wlt-cortex) [inline cvm/quantization/from_tensorflow.py:646]: Why this exception of TypeError? can you explain a lot more via comments?

File: cvm/quantization/from_tensorflow.py [hunk 3]
```diff
@@ -231,7 +261,8 @@ def _impl(inputs, attrs, params):
 
         if pool_type == "avg":
             # TODO: this actually should be count_include_pad False
-            pool_attr['count_include_pad'] = True
+            # pool_attr['count_include_pad'] = True
+            pool_attr['count_include_pad'] = False
```
Comment (wlt-cortex) [inline cvm/quantization/from_tensorflow.py]: This value actually should be false, and then why set it true before?

File: cvm/quantization/dataset.py [hunk 4]
```diff
@@ -16,7 +16,7 @@
 import logging
 
 dataset_dir = path.expanduser("~/.cvm")
-src = "http://192.168.50.210:8827"
+src = "http://0.0.0.0:8827"
```
Comment (wlt-cortex) [inline cvm/quantization/dataset.py:19]: `src` is set as 0.0.0.0, which seems like not a valid URL. Does it really work?

File: cvm/quantization/tfm_ops.py [hunk 5]
```diff
@@ -1189,10 +1190,10 @@ def quantize(self, op, **kwargs):
         # scales[out.attr('name')] = scales[X_name]
         # return out
 
-        precs[name][OUT_KEY] = precs[X.attr('name')][OUT_KEY]
         scales[name] = iscale = scales[X.attr('name')]
         a_min = sutils.get_attr(attrs, "a_min") * iscale
         a_max = sutils.get_attr(attrs, "a_max") * iscale
+        precs[name][OUT_KEY] = min(get_bit(th_dict[name]*iscale), get_bit(int(a_max)))
```
Comment (wlt-cortex) [inline cvm/quantization/tfm_ops.py:1196]: How the tight precision influence the quantized model? Is it necessary, just for confusion?

File: cvm/quantization/from_tensorflow.py (hunk 1 above)
Comment (declmal) [reply cvm/quantization/from_tensorflow.py]: already fixed

File: cvm/quantization/from_tensorflow.py (hunk 3 above)
Comment (declmal) [reply cvm/quantization/from_tensorflow.py]: When I first see here, it was set to True, so I changed this to False as implied in the comment.

File: cvm/quantization/from_tensorflow.py (hunk 2 above)
Comment (declmal) [reply cvm/quantization/from_tensorflow.py:646]: Ususally in tensorflow models, attr['dtype'].type will get a value of 1, which represents float32 in tensorflow. 
However in resnet_v2_101, attr['dtype'].type will get a value of 0, and TypeError occurs. set the value to default_tf_dtype would work fine.

File: cvm/quantization/dataset.py (hunk 4 above)
Comment (declmal) [reply cvm/quantization/dataset.py:19]: The original src was found invalid when I was tring to dowload the mnist dataset. After revising the address, the dateset could be successfully downloaed.

File: cvm/quantization/tfm_ops.py (hunk 5 above)
Comment (declmal) [reply cvm/quantization/tfm_ops.py:1196]: The original precs for `clip` is not the tight one. After restoring all the `convolution` and `clip`, the accuracy looks good. Howver, while using `clip` quantization and keeping `convolution` restored, the accuracy was poor.
When clip was switched to tight precision, the accuracy works well for all the lite models.

Comment (declmal) [reply cvm/quantization/tfm_ops.py]: mobilenetv2_1.0 works well no matter `clip` is restored or not, which is strange.

Comment (declmal) [reply cvm/quantization/tfm_ops.py:1196]: mobilenetv2_1.0 works well no matter prec out_key for `clip` is tight or not, which is strange.

Comment (declmal) [reply cvm/quantization/tfm_ops.py:1196]: And clip only exist in mobilenetv2_1.0 and lite models

File: cvm/quantization/from_tensorflow.py (hunk 2 above)
Comment (wlt-cortex) [reply cvm/quantization/from_tensorflow.py:646]: Then what the value of 0 means, do you know?

File: cvm/quantization/dataset.py (hunk 4 above)
Comment (wlt-cortex) [reply cvm/quantization/dataset.py:19]: 0.0.0.0? It's strange, try to set 127.0.0.1? Maybe the 0.0.0.0 indicates localhost?

File: cvm/quantization/tfm_ops.py (hunk 5 above)
Comment (wlt-cortex) [reply cvm/quantization/tfm_ops.py:1196]: That's cause op clip is converted from ReLu6, mobilenetv2 use this operator in its network, and many lite models seem to prefer the ReLu6 op, don't know and just guess it. As you point out, the clip op that is necessary for tight or not relates to the requant&restore pass for post-quantization accuracy promotion, is it? I have created a new card about the post-quantization accuracy promotion in the [Trello](https://trello.com/c/gxsqqFZ6/3-accuracy-finetune-via-layers-restore) you'd better check out.

File: cvm/quantization/from_tensorflow.py (hunk 2 above)
Comment (declmal) [reply cvm/quantization/from_tensorflow.py:646]: 0 means `DT_INVALID`

Usage:
indicate a DataType field has not been set.

Declaration:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/types.proto

Which models have:
ssd_mobilenet_v2_coco, resnet_v2_101_lite
the first two layers of both models are repectively: FIFOQueueV2, QueueDequeueManyV2
I don't know how to realize this two layer in mxnet.

My solution:
fuse first two layer into an input layer, and directly set the data type as DT_FLOAT which is 1

Comment (declmal) [reply cvm/quantization/from_tensorflow.py:646]: Correctness: only resnet_v2_101_lite, ssd_mobilenet_v2_coco is not.

File: cvm/quantization/from_tensorflow.py (hunk 3 above)
Comment (wlt-cortex) [reply cvm/quantization/from_tensorflow.py]: Aha, I remembered it. The AvgPooling convention from TensorFlow refers to the transformation [source code](https://github.com/apache/incubator-tvm/blob/master/python/tvm/relay/frontend/tensorflow.py#L240) from TF into TVM's relay IR. And I changed it into True directly for the purpose that our MRT only supports count_include_pad = True at the convention design. May models that go under from_tensorflow pass not accepted by out MRT after changing it to False, that's a problem, and you can check out the models containing AvgPooling operator such as resnet_v1?

And another point is that whereas the True value is set in the relay IR transformation, the default process logic in TensorFlow is indefinite to us. We'd better research the real source code logic in TF and then examine the count_include_pad's value.

The current suggestion is back to the True default value and adds more comments in the source such as TODO.

File: cvm/quantization/from_tensorflow.py (hunk 2 above)
Comment (wlt-cortex) [reply cvm/quantization/from_tensorflow.py:646]: En, It seems the operators: FIFOQueueV2, QueueDequeueManyV2 that brings into DT_INVALID problem. I'll try to solve it when I'm free and this PR is stuck here temporarily.

Comment (wlt-cortex) [reply cvm/quantization/from_tensorflow.py:646]: @declmal , Add TODO in the source code at here, I'll merge the PR 57 first before [PR 58](https://github.com/CortexFoundation/tvm-cvm/pull/58)
